src/CodeAnalysis/CSharp/MetaAnalyzers/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationDescription), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+
+        public static DiagnosticDescriptor DoNotIgnoreReturnValueDiagnosticRule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocation,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.AnalyzerCorrectness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(DoNotIgnoreReturnValueDiagnosticRule);
+            }
+        }
+
+        private static readonly string s_solutionFullName = @"Microsoft.CodeAnalysis.Solution";
+        private static readonly string s_projectFullName = @"Microsoft.CodeAnalysis.Project";
+        private static readonly string s_documentFullName = @"Microsoft.CodeAnalysis.Document";
+        private static readonly string s_syntaxNodeFullName = @"Microsoft.CodeAnalysis.SyntaxNode";
+
+        // A list of known immutable object names
+        private static readonly List<string> s_immutableObjectNames = new List<string>()
+        {
+            s_solutionFullName,
+            s_projectFullName,
+            s_documentFullName,
+            s_syntaxNodeFullName,
+        };
+
+        private static readonly string s_Add = "Add";
+        private static readonly string s_Remove = "Remove";
+        private static readonly string s_Replace = "Replace";
+        private static readonly string s_With = "With";
+
+        private static readonly List<string> s_immutableMethodNames = new List<string>()
+        {
+            s_Add,
+            s_Remove,
+            s_Replace,
+            s_With,
+        };
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCodeBlockAction(AnalyzeCodeBlockForIgnoredReturnValues);
+        }
+
+        public void AnalyzeCodeBlockForIgnoredReturnValues(CodeBlockAnalysisContext context)
+        {
+            var model = context.SemanticModel;
+            var syntaxNode = context.CodeBlock;
+
+            //Find invocations that ignore the return value
+            //We're looking for invocations that are not within an equals value clause or a return statement.
+            var candidateInvocations = syntaxNode.DescendantNodes()
+                .OfType<InvocationExpressionSyntax>()
+                .Where(n => n.Parent is StatementSyntax && !(n.Parent is ReturnStatementSyntax));
```

Comment:
Registering a code block analyzer that walks all descendant nodes to do analysis is not recommended. I would recommend the following principles:

1. If you want to report diagnostics that belong at code block level, do it via a CodeBlock action. For an example, look at this [sample](https://github.com/dotnet/roslyn/blob/master/src/Samples/CSharp/Analyzers/CSharpAnalyzers/CSharpAnalyzers/StatelessAnalyzers/CodeBlockAnalyzer.cs). You shouldn't be walking a descendants syntax nodes of a code block in such an action.

2. If you are reporting diagnostics for a syntax node that are not dependent on any other nodes within the tree (pure syntax diagnostics), you should register syntax node actions, example [here](https://github.com/dotnet/roslyn/blob/master/src/Samples/CSharp/Analyzers/CSharpAnalyzers/CSharpAnalyzers/StatelessAnalyzers/SyntaxNodeAnalyzer.cs)

3. If you are writing a stateful analyzer where you need to perform analysis across a code block, compilation, etc., then you should use the stateful actions (start/end pair). See example [here](https://github.com/dotnet/roslyn/blob/master/src/Samples/CSharp/Analyzers/CSharpAnalyzers/CSharpAnalyzers/StatefulAnalyzers/CodeBlockStartedAnalyzer.cs)

Note that the analyzer driver ensures that syntax node actions for nodes within executable blocks are always invoked after semantic analysis/binding, so the semantic model given to the analyzer will have cached bound nodes regardless of where you register the syntax node action.